### ISSUE TITLE
Re-render breaks onOpenEnd and onCloseStart

### ISSUE THREAD

**Issue by adoniscoder** [osdnk/react-native-reanimated-bottom-sheet#183]
Hi, I have callbacks for onOpenEnd and onCloseStart that will cause to state change on the parent component but the problem is after the first call these two methods will not be called anymore, if I remove the setState it will work as normal.

**Comment by adoniscoder** [comment]
`
   import * as React from 'react';
   import {Dimensions, StyleSheet, TouchableWithoutFeedback} from 'react-native';
   import BottomSheet from 'reanimated-bottom-sheet';
   import Animated from 'react-native-reanimated';
   import {ActionSheetBloc} from 'bloc';

   interface Props {
      children: React.ReactNode;
       title: string;
   }

   const {height} = Dimensions.get('window');

   class ActionSheet extends React.Component<Props> {
        private bs = React.createRef<BottomSheet>();
       private fall = new Animated.Value(1);
       private animatedHeaderContentOpacity = Animated.interpolate(this.fall, {
            inputRange: [0, 0.25, 1],
            outputRange: [1, 0.25, 0],
           extrapolate: Animated.Extrapolate.CLAMP,
       });
       private snapPoints = [height - 50, 300, 50];
       private emitter = new ActionSheetBloc();

    onHeaderPress = (): void => {
        if (this.bs && this.bs.current) {
            this.bs.current.snapTo(1);
        }
    };

    sheetOpenEnd = (): void => {
        this.notifyStateChanged(1);
    };

    sheetCloseStart = (): void => {
        this.notifyStateChanged(0);
    };

    notifyStateChanged = (state: number): void => {
        this.emitter.stateChanged(this.props.title, state);
    };

    renderHeader = (): React.ReactNode => {
        const animatedBackgroundOpacity = Animated.sub(1, this.animatedHeaderContentOpacity);
        return (
            <TouchableWithoutFeedback key={'header-container'} onPress={this.onHeaderPress}>
                <Animated.View style={styles.headerContainer}>
                    <Animated.View
                        style={[
                            styles.headerBackground,
                            {
                                opacity: animatedBackgroundOpacity,
                            },
                        ]}></Animated.View>
                </Animated.View>
            </TouchableWithoutFeedback>
        );
    };

    render(): React.ReactNode {
        return (
            <React.Fragment>
                <BottomSheet
                    ref={this.bs}
                    initialSnap={1}
                    enabledBottomInitialAnimation
                    onOpenEnd={this.sheetOpenEnd}
                    onCloseStart={this.sheetCloseStart}
                    callbackNode={this.fall}
                    snapPoints={this.snapPoints}
                    renderContent={(): React.ReactNode => this.props.children}
                    renderHeader={this.renderHeader}
                />
            </React.Fragment>
        );
    }
}

const styles = StyleSheet.create({
    headerContainer: {
        backgroundColor: 'white',
        alignItems: 'center',
        justifyContent: 'center',
        paddingVertical: 17,
    },
    headerBackground: {
        // ...StyleSheet.absoluteFillObject,
        width: 135,
        height: 9,
        borderWidth: 1,
        borderColor: '#979797',
        borderRadius: 5,
        // overflow: 'hidden',
        alignSelf: 'center',
        backgroundColor: '#6D7278',
        // zIndex: 9,
    },
    headerHandler: {
        width: 135,
        height: 9,
        backgroundColor: '#6D7278',
        borderColor: '#979797',
        borderRadius: 10,
        borderWidth: 1,
    },
    handlerContainer: {
        position: 'absolute',
        alignSelf: 'center',
        top: 10,
        height: 20,
        width: 20,
    },
    handlerBar: {
        position: 'absolute',
        backgroundColor: 'red',
        top: 5,
        borderRadius: 3,
        height: 5,
        width: 20,
    },
});

export default ActionSheet;
`

**Comment by stachu2k** [comment]
I have the same issue. For me onOpenStart nad onOpenEnd events are not always firing after upgrade to react-native 0.61.5 and react-native-navigation: 3.7.0. I have the latest 1.0.0-alpha.19 version of bottom sheet.

**Comment by wassgha** [comment]
Same here.

**Comment by stachu2k** [comment]
Any news?

**Comment by adoniscoder** [comment]
if you pass all the events to the bottomsheet it works well although it isn't a clean approach but for now it fixes the problem.
`
    <BottomSheet
                    ref={this.bs}
                    initialSnap={1}
                    enabledBottomInitialAnimation
                    onOpenStart={(): void => {}}
                    onOpenEnd={this.sheetOpenEnd}
                    onCloseStart={this.sheetCloseStart}
                    onCloseEnd={(): void => {}}
                    callbackNode={this.fall}
                    snapPoints={this.snapPoints}
                    renderContent={(): React.ReactNode => this.props.children}
                    renderHeader={this.renderHeader}
                />
`

**Comment by stachu2k** [comment]
@adoniscoder Sadly, it doesn't work for me :(

**Comment by karimcambridge** [comment]
`onCloseEnd` really needs a fix.

**Comment by yzalvov** [comment]
@stachu2k, if it's still relevant try using the patched version I posted in #136 . It puts all my open/close events in order. The code is also available via the gist I created, [ReanimatedBottomSheet_patched.js](https://gist.github.com/yzalvov/28133b3654ddb2724b201902937608cd).

**Comment by karimcambridge** [comment]
> @stachu2k, if it's still relevant try using the patched version I posted in #136 . It puts all my open/close events in order. The code is also available via this gist [ReanimatedBottomSheet_patched.js](https://gist.github.com/yzalvov/28133b3654ddb2724b201902937608cd).

Can you create a pull request?

**Comment by yzalvov** [comment]
@karimcambridge I’m glad it helps. 

As for the pull request, I assume that a PR usually is a fix or an addition to the original code. I can see the part of the original code for the event handlers and would love to PR a fix, but as I said before I’m not sure I’d be able to make enough time at the moment for digging into it so that other things won’t be broken.

Also I use hooks which, as I guess, is not what the original code based on. Another complication. 

So a small nice and easy patch wrapper looks like a way to go.

**Comment by yzalvov** [comment]
@karimcambridge , @osdnk I think for the fix it's safe to use `Animated` conditions from the patch I posted. I guess this part may be tricky so the sketch I helped myself with may be useful.

![20200530-bottom-sheet-events-fix](https://user-images.githubusercontent.com/25827414/83364189-2a6f6c80-a3a8-11ea-8529-4abef73e1343.png)

**Comment by mayconmesquita** [comment]
@yzalvov using the patch solved the issue here! Thanks!

**Comment by jerearaujo03** [comment]
The issue is not solved yet!

**Comment by yzalvov** [comment]
Hi @jerearaujo03 ,
I guess you've managed to pass props to my patched version [ReanimatedBottomSheet_patched.js](https://gist.github.com/yzalvov/28133b3654ddb2724b201902937608cd)?

**Comment by jerearaujo03** [comment]
Yes! It works well (maybe there is a small delay when executing `onOpenStart `), but it works

**Comment by yzalvov** [comment]
Hi @jerearaujo03 

You may try to pass in `callbackThreshold` prop as less than 0.01 (which is the default value) and check if `onOpenStart` fires sooner. 

For more details consult line 36 of the patch. The `bottomTippingPoint` is defined there.

**Comment by jerearaujo03** [comment]
Yes, i've already tried but it doesn't worked. Likewise it works very well

**Comment by danieltovesson** [comment]
@yzalvov your patch almost works for me. There is an issue when you swipe very quickly and for some reason, only `handleOpenEnd` is called then, so not `handleOpenStart`. I fixed this by adding
```
function handleOpenEnd() {
    isCloseEndAllowed.current = true;
    if (onOpenEnd) onOpenEnd();
}
```
below the `handleOpenStart` function and changed the line `() => onOpenEnd && onOpenEnd()` to `handleOpenEnd`.

Let me know if that makes sense

**Comment by yzalvov** [comment]
@danieltovesson hi there. It'd make sense only if it'd work. I guess it doesn't. Because nothing changed in your update, except you wrapped the `onOpenEnd` function with a new `handleOpenEnd` one.

But if somehow it works for you, I'd just remove the `isCloseEndAllowed.current = true;` line from your code as it is already there in my code and doesn't have to be repeated.

**Comment by victoriomolina** [comment]
Still having this issue :C

**Comment by victoriomolina** [comment]
Btw, I have found a tricky solution to this error.

```
import React, { forwardRef, useImperativeHandle, useRef } from "react";
import { View, Text, TouchableHighlight } from "react-native";
import Animated, { sub } from "react-native-reanimated";
import ReanimatedBottomSheet from "reanimated-bottom-sheet";
import { Divider, useTheme } from "react-native-paper";
import PropTypes from "prop-types";

import globalStyles from "../../styles/globalStyles";
import styles from "./styles/styles";

const BottomSheet = forwardRef(
  (
    {
      topItems = [],
      bottomItems = [],
      sheetHeight = 0,
      initialSnap = 1,
      onOpenStart,
      onOpenEnd,
      onCloseStart,
      onCloseEnd,
    },
    ref
  ) => {
    useImperativeHandle(ref, () => ({
      open,
      close,
    }));

    const { colors } = useTheme();

    const bottomSheetRef = useRef(null);
    const fall = useRef(new Animated.Value(1)).current;
    const isCloseEndAllowed = useRef(false); // 🐛 BUG - https://github.com/osdnk/react-native-reanimated-bottom-sheet/issues/136#issuecomment-565748783

    const open = () => {
      bottomSheetRef.current.snapTo(0);
    };

    const close = () => {
      bottomSheetRef.current.snapTo(1);
    };

    const handleOnOpenStart = () => {
      isCloseEndAllowed.current = true;
      onOpenStart?.();
    };

    const handleOnCloseEnd = () => {
      if (!isCloseEndAllowed.current) return;
      onCloseEnd?.();
    };

    const renderContent = () => {
      return (
        <View style={styles.contentContainer}>
          <View
            style={[styles.buttonsContainer, { backgroundColor: colors.white }]}
          >
            {topItems.map((item, index) => {
              const { disabled, title, titleColor, onPress } = item;

              return (
                <React.Fragment key={index}>
                  <TouchableHighlight
                    disabled={disabled}
                    underlayColor="#EBEBEB"
                    onPress={onPress}
                    style={[
                      styles.button,
                      index === 0 && {
                        borderTopLeftRadius: 15,
                        borderTopRightRadius: 15,
                      },
                      index === topItems.length - 1 && {
                        borderBottomLeftRadius: 15,
                        borderBottomRightRadius: 15,
                      },
                    ]}
                  >
                    <Text
                      style={[
                        styles.topButtonText,
                        {
                          color: !disabled
                            ? titleColor || colors.azureRadiance
                            : colors.disabled,
                        },
                      ]}
                    >
                      {title}
                    </Text>
                  </TouchableHighlight>

                  {index !== topItems.length - 1 && <Divider />}
                </React.Fragment>
              );
            })}
          </View>

          <View
            style={[styles.buttonsContainer, { backgroundColor: colors.white }]}
          >
            {bottomItems.map((item, index) => {
              const { disabled, title, titleColor, onPress } = item;

              return (
                <React.Fragment key={index}>
                  <TouchableHighlight
                    disabled={disabled}
                    underlayColor="#EBEBEB"
                    onPress={onPress}
                    style={[
                      styles.button,
                      index === 0 && {
                        borderTopLeftRadius: 15,
                        borderTopRightRadius: 15,
                      },
                      index === bottomItems.length - 1 && {
                        borderBottomLeftRadius: 15,
                        borderBottomRightRadius: 15,
                      },
                    ]}
                  >
                    <Text
                      style={[
                        styles.bottomButtonText,
                        {
                          color: !disabled
                            ? titleColor || colors.azureRadiance
                            : colors.disabled,
                        },
                      ]}
                    >
                      {title}
                    </Text>
                  </TouchableHighlight>

                  {index !== bottomItems.length - 1 && <Divider />}
                </React.Fragment>
              );
            })}
          </View>
        </View>
      );
    };

    /*
      🐛 BUG - Re-renders break BottomSheet's events. 

      With the help of abstraction, I have come up with a hacky solution to the problem. This code looks
      ugly and is not optimal, but being something related to a dependency, it is the only thing I can do.
    
      https://github.com/osdnk/react-native-reanimated-bottom-sheet/issues/183
    */

    return (
      <View
        style={[globalStyles.overlayed, styles.container]}
        pointerEvents={!sheetHeight ? "none" : "box-none"} // tricky
      >
        <ReanimatedBottomSheet
          ref={bottomSheetRef}
          snapPoints={[sheetHeight, "-10%"]} // tricky
          initialSnap={initialSnap}
          renderContent={renderContent}
          callbackNode={fall}
          enabledContentTapInteraction={false}
          onOpenStart={handleOnOpenStart}
          onOpenEnd={onOpenEnd}
          onCloseStart={onCloseStart}
          onCloseEnd={handleOnCloseEnd}
        />

        {/* Overlay Shadow */}
        <Animated.View
          pointerEvents={"box-only"}
          style={[
            globalStyles.overlayed,
            {
              opacity: sub(0.5, fall),
              backgroundColor: colors.black,
            },
          ]}
        />
      </View>
    );
  }
);

BottomSheet.propTypes = {
  topItems: PropTypes.arrayOf(
    PropTypes.shape({
      disabled: PropTypes.bool,
      title: PropTypes.string.isRequired,
      titleColor: PropTypes.string,
      onPress: PropTypes.func,
    })
  ),
  bottomItems: PropTypes.arrayOf(
    PropTypes.shape({
      disabled: PropTypes.bool,
      title: PropTypes.string.isRequired,
      titleColor: PropTypes.string,
      onPress: PropTypes.func,
    })
  ),
  sheetHeight: PropTypes.number,
  initialSnap: PropTypes.number,
  onOpenStart: PropTypes.func,
  onOpenEnd: PropTypes.func,
  onCloseStart: PropTypes.func,
  onCloseEnd: PropTypes.func,
};

export default BottomSheet;

```
